### ISSUE TITLE
Improving the support of Copilot Chat /workspace agent for selected symbols

### ISSUE THREAD

**Issue by lukka** [microsoft/vscode-copilot-release#780]
The Copilot Chat has introduced the `/workspace` agent [that is interfacing with the language service (LSP based)](https://github.blog/changelog/2023-11-30-github-copilot-november-30th-update/#improved-explanation-context-in-copilot-chat-in-vs-code) of the codebase. While the feature is working good with e.g., Typescript/Javascript language, it looks like the integration for C++ code bases could be further improved. 

In my experiments, when preparing the context of the prompt presented to Copilot the C++ LSP provided by the [vscode-cpptools](https://github.com/microsoft/vscode-cpptools) extension is not being queried for declarations, for definition or references (e.g., find-all-refs) across the code base. The declaration/definition/references could likely be useful to enrich the prompt with additional context and lead to an increasing quality of the answers.

This is evident when the Copilot Chat's `/workspace` agent explicitly states that "it can't provide information on them without more details about your workspace.” or that "Without any additional code, it's hard to provide a detailed explanation".

@isidorn this issue is posted by the C++ crew who contacted you via email about improving the Copilot Chat UX for C++ users.

**Comment by mjbvz** [comment]
There's nothing super obvious that would cause a C++ support to be worse than TS. `@workspace` has semantic understanding of C++ and `@workspace` uses language IntelliSense features in a language agnostic way (we currently only use symbols, not declarations, definition, or references)

A few related things we can investigate tho:

- When we break a C++ file into chunks, are we doing it in the right way or is some content being lost?
- Does the C++ extension return IntelliSense in an unexpected format?

---

@lukka Can you please share an example project plus some example `@workspace `question that don't work well

**Comment by lukka** [comment]
@mjbvz I shared with you a document which I already had and that shows some examples. I can also post them in here if desired.

**Comment by lukka** [comment]
@mjbvz let me know if you could access the document I shared with you via email. I add here below another example, where the header file of the `cmTextConfigureCommand` class is not being used as a reference (only the file where the selection is located is), and the answer complains exactly about that: 

> The active selection cmCTestConfigureCommand appears to be a class or function name in C++. However, without additional context, it's difficult to provide a detailed explanation.

and it continues saying that "If cmCTestConfigureCommand  is a class", and then "If cmCTestConfigureCommand is a function", which I believe it should have been clear if the header was included as a reference.

<img width="598" alt="image" src="https://github.com/microsoft/vscode-copilot-release/assets/681992/defa39c9-db6c-4868-b571-dedc4cf968f4">


<br>
<br>

I am willing help and debug the scenario of using the [cpp-tools VS Code C++ language service](https://github.com/microsoft/vscode-cpptools) with Copilot Chat. Regarding your questions:

> (we currently only use symbols, not declarations, definition, or references)

I may be doing something incorrectly, but I cannot get any breakpoint hit at [WorkspaceSymbolProvider.providerWorkspaceSymbols](https://github.com/microsoft/vscode-cpptools/blob/3f1e851152626e9af01387b672d565bf25498990/Extension/src/LanguageServer/Providers/workspaceSymbolProvider.ts#L16). Any clue or hint? Is it cached in some file and I need to wipe it out?

> When we break a C++ file into chunks, are we doing it in the right way or is some content being lost?
> Does the C++ extension return IntelliSense in an unexpected format?

I would be happy to help to answer the following questions, but I'd need more info about which code to look into (and where to put breakpoints basically).

**Comment by lukka** [comment]
@sbatten @isidorn @mjbvz  I can report that on latest VS Insiders (`ec291c1268`) and latest vscode-copilot built on `main` (at `8246dab7e530d5`) is still exhibiting the same behavior described above, where the [provideWorkspaceSymbol](https://github.com/microsoft/vscode-cpptools/blob/aa5601041c9d9f6cc6bb2492cd69c52af7db5992/Extension/src/LanguageServer/Providers/workspaceSymbolProvider.ts#L16) is not queried by vscopilot-chat as I would expect, and the C++ language service is then not leveraged to pick the right  references used to craft the prompt and improve the answer accuracy.

I'd be super happy to help debugging this scenario, but I need more information about how to proceed, that is:

> When we break a C++ file into chunks, are we doing it in the right way or is some content being lost?
> Does the C++ extension return IntelliSense in an unexpected format?
> 

I would be happy to help to answer the following questions, but I'd need more info about which code to look into (and where to put breakpoints basically).

**Comment by mjbvz** [comment]
I just `@workspace` in a basic c++ project (no `/explain`). Seems to work fine and also I see the calls to `getWorkspaceSymbols`

![Image](https://github.com/microsoft/vscode-copilot-release/assets/12821956/a0f84e6d-5bfb-4b00-b18f-7d9f3769fa35)


![Image](https://github.com/microsoft/vscode-copilot-release/assets/12821956/f62f95ea-288c-4e2c-91bb-f9e15a410716)


@lukka `/explain` works differently and is also implemented differently. It will not make calls to the language service at the moment. I'll add in the owners on our side but please share an example project too to help with investigation

**Comment by joyceerhl** [comment]
`/explain` does make calls to the language service to resolve functions, classes and types in the current selection, but it currently expects a selection of code rather than just a symbol. This milestone we are working on improving its behavior for when the cursor happens to be on a symbol, or a symbol is selected and nothing else.

**Comment by lukka** [comment]
@joyceerhl thank you for explaining that selecting a symbol is a known issue when using `@workspace /explain` and that it will be addressed.

@mjbvz thank you for pointing out that the WorkspaceSymbolProvider is used by `@workspace` (but not by `@workspace /explain` indeed).

Based on your input, I experimented more with Copilot Chat and  C++ code, and here below I have more feedback to provide.

## **TL DR:**
Here the two issues I noticed:
- @joyceerhl I can't see any interaction with the C++ language service when using the `@workspace /explain` subcommand, and only one reference is ever used in the question, the one where the selection occurs.
- @mjbvz [as you asked previously](https://github.com/microsoft/vscode-copilot-release/issues/780#issuecomment-1908637804), while `@workspace` is doing a great job at collecting references, it looks like the references content is improperly split.

_Please feel free to ping me anytime to help with repro step or with debugging anything on the C++ Lang Service side._

## Here the details

I played with the two questions below and I have identified two potential issues, here the steps to repro:
- I used installed prereleases of [CMake Tools](https://github.com/microsoft/vscode-cmake-tools), [C++](https://github.com/microsoft/vscode-cpptools), CopilotChat `v0.13.2024020601` and VSCode `1.87-insider-2024-02-06` (`cf7ddbb51`).
- Checkout the [CMake code base](https://github.com/Kitware/CMake), configure with CMake and build it
- then I asked the following two questions below.


### Question: `@workspace /explain what is ComputeDefines doing?` 

@joyceerhl in the next screenshot there are two questions that, to being properly answered, they need the context from other files than the one where the selection is done: the problem is that I do not see Copilot Chat leveraging the C++ Language Service for finding references of the `ComputeDefines` member function, which would have been fundamental to answer both questions:

![image](https://github.com/microsoft/vscode-copilot-release/assets/681992/bf1552ed-0dd0-4d68-af92-240e43becb2c)

The only reference used is the selection I've done at `Source\cmNinjaTargetGenerator.h` lines 99:102, and I would expect to see at least the file `Source\cmNinjaTargetGenerator.cxx` referenced, which would be needed to answer the second question.

### Question: `@workspace explain what is ComputeDefines doing?`

@mjbvz  in the next screenshot, the answer to the question is very well done, especially because there are 8 references being collected and used, especially the file `Source\cmNinjaTargetGenerator.cxx` which contains the key to a proper answer.

![image](https://github.com/microsoft/vscode-copilot-release/assets/681992/d5b3d4e7-4f22-4063-8a5a-0589c3b6c04b)

As you [asked before](https://github.com/microsoft/vscode-copilot-release/issues/780#issuecomment-1908637804), the portion of text from each reference is not as I would expect, going up to the end of a function definition for example, but they are just terminating in the middle of anything.

For example:

- This is just a blank line:
    ![image](https://github.com/microsoft/vscode-copilot-release/assets/681992/9d83a944-34e0-4271-89c5-10519a0bf5d5)

- These are not going up to the end of the function body:
    ![image](https://github.com/microsoft/vscode-copilot-release/assets/681992/7086c9a4-2b10-4db2-a3e6-8c1c7f2ef654)
    ![image](https://github.com/microsoft/vscode-copilot-release/assets/681992/5d9596d0-3ba0-4644-b171-cf9c3a82aed7)
    ![image](https://github.com/microsoft/vscode-copilot-release/assets/681992/b800da08-c952-47a4-9004-eba0cc3ac62f)

**Comment by joyceerhl** [comment]
Duplicating this issue against https://github.com/microsoft/vscode-copilot-release/issues/868 for `/explain` which should be fixed in our pre-release.
